Modules/_testcapi/heaptype_relative.c

```diff
@@ -0,0 +1,289 @@
+#define Py_LIMITED_API 0x030c0000 // 3.12
+#include "parts.h"
+#include <stddef.h>               // max_align_t
+#include <string.h>               // memset
+
+#ifdef LIMITED_API_AVAILABLE
+
+static PyType_Slot empty_slots[] = {
+    {0, NULL},
+};
+
+static PyObject *
+make_sized_heaptypes(PyObject *module, PyObject *args)
+{
+    PyObject *base = NULL;
+    PyObject *sub = NULL;
+    PyObject *instance = NULL;
+    PyObject *result = NULL;
+
+    int extra_base_size, basicsize;
+
+    int r = PyArg_ParseTuple(args, "ii", &extra_base_size, &basicsize);
+    if (!r) {
+        goto finally;
+    }
+
+    PyType_Spec base_spec = {
+        .name = "_testcapi.Base",
+        .basicsize = sizeof(PyObject) + extra_base_size,
+        .flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE,
+        .slots = empty_slots,
+    };
+    PyType_Spec sub_spec = {
+        .name = "_testcapi.Sub",
+        .basicsize = basicsize,
+        .flags = Py_TPFLAGS_DEFAULT,
+        .slots = empty_slots,
+    };
+
+    base = PyType_FromMetaclass(NULL, module, &base_spec, NULL);
+    if (!base) {
+        goto finally;
+    }
+    sub = PyType_FromMetaclass(NULL, module, &sub_spec, base);
+    if (!sub) {
+        goto finally;
+    }
+    instance = PyObject_CallNoArgs(sub);
+    if (!instance) {
+        goto finally;
+    }
+    char *data_ptr = PyObject_GetTypeData(instance, (PyTypeObject *)sub);
+    if (!data_ptr) {
+        goto finally;
+    }
+    Py_ssize_t data_size = PyType_GetTypeDataSize((PyTypeObject *)sub);
+    if (data_size < 0) {
+        goto finally;
+    }
```

Comment:
`PyType_GetTypeDataSize` cannot return negative values; the check is not needed.

Reply:
Ditto for the usages in the rest of the file.

Reply:
Here I want to test with intended usage, rather than rely on current details.

Reply:
Fair enough